src/errors.rs

```diff
@@ -10,127 +10,245 @@
+    #[fail(display = "Stale epoch. {}", message)]
+    StaleEpoch { message: String },
+    /// No region is found for the given id.
+    #[fail(display = "Region {} is not found. {}", region_id, message)]
+    RegionNotFound { region_id: u64, message: String },
+    /// Invalid key range to scan. Only left bounded intervals are supported.
+    #[fail(display = "Only left bounded intervals are supported")]
+    InvalidKeyRange,
+    /// No such key
+    #[fail(display = "Key does not exist")]
+    NoSuchKey,
+    /// Cannot set an empty value
+    #[fail(display = "Cannot set an empty value")]
+    EmptyValue,
+    /// Scan limit exceeds the maximum
+    #[fail(display = "Limit {} exceeds max scan limit {}", limit, max_limit)]
+    MaxScanLimitExceeded { limit: u32, max_limit: u32 },
+    /// Wraps `kvproto::kvrpcpb::KeyError`
+    #[fail(display = "{:?}", _0)]
+    KeyError(kvproto::kvrpcpb::KeyError),
+    /// A string error returned by TiKV server
+    #[fail(display = "Kv error. {}", message)]
+    KvError { message: String },
+    /// Reconstructed `kvproto::errorpb::KeyNotInRegion`
+    #[fail(
+        display = "Key {:?} is not in region {}: [{:?}, {:?})",
+        key, region_id, start_key, end_key
+    )]
+    KeyNotInRegion {
+        key: Vec<u8>,
+        region_id: u64,
+        start_key: Vec<u8>,
+        end_key: Vec<u8>,
+    },
+    /// Reconstructed `kvproto::errorpb::ServerIsBusy`
+    #[fail(display = "Server is busy: {}. Backoff {} ms", reason, backoff_ms)]
+    ServerIsBusy { reason: String, backoff_ms: u64 },
+    /// Represents `kvproto::errorpb::StaleCommand` with additional error message
+    #[fail(display = "Stale command. {}", message)]
+    StaleCommand { message: String },
+    /// Represents `kvproto::errorpb::StoreNotMatch` with additional error message
+    #[fail(
+        display = "Requesting store {} when actual store is {}. {}",
+        request_store_id, actual_store_id, message
+    )]
+    StoreNotMatch {
+        request_store_id: u64,
+        actual_store_id: u64,
+        message: String,
+    },
+    /// Represents `kvproto::errorpb::RaftEntryTooLarge` with additional error message
+    #[fail(
+        display = "{} bytes raft entry of region {} is too large. {}",
+        entry_size, region_id, message
+    )]
+    RaftEntryTooLarge {
+        region_id: u64,
+        entry_size: u64,
+        message: String,
+    },
+    #[fail(display = "{}", message)]
+    InternalError { message: String },
+}
+
+impl Fail for Error {
+    fn cause(&self) -> Option<&Fail> {
+        self.inner.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.inner.backtrace()
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Display::fmt(&self.inner, f)
+    }
+}
+
+impl Error {
+    pub fn kind(&self) -> &ErrorKind {
+        self.inner.get_context()
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn io(inner: std::io::Error) -> Self {
```

Comment:
Hm, maybe we can do `impl From<std::io::Error> for Error` (etc) instead?

Reply:
Implemented `From` for types which have a direct wrapper.